modules/gapi/include/opencv2/gapi/cpu/gcpukernel.hpp

```diff
@@ -127,16 +128,28 @@ class GAPI_EXPORTS GCPUContext
 class GAPI_EXPORTS GCPUKernel
 {
 public:
-    // This function is kernel's execution entry point (does the processing work)
-    using F = std::function<void(GCPUContext &)>;
+    // This function is a kernel's execution entry point (does the processing work)
+    using RunF = std::function<void(GCPUContext &)>;
+    // FIXME: Workaround to hide dangerous bug:
+    //        If we add callbacks here as std::function-s then we will get
+    //        into segfault with "stack smashing detected" error for GPU tests.
+    //        Segfault will be happened on GOCLBackendImpl::unpackKernel exit.
+    //        But, if we comment last statement in GOCLBackendImpl::unpackKernel
+    //        so this issue will not be reproducible.
+    // This function is a stateful kernel's setup routine (configures state)
+    using SetupF = void(*)(GMetaArgs, GArgs, GArg &);
+    // This function is a stateful kernel's execution entry point (does the processing work)
+    using StRunF = void(*)(GCPUContext &, GArg &);
```

Comment:
Those should remain be std::functions. The problem is somewhere else. Also, the SetupF should take its `GMetaArgs` and `GArgs` as `const &`.

Reply:
Also, please drop a dedicated StRunF. Why use it at all? `GCPUContext` should be enough.

Let me say it this way. The main reason to wrap everything into a structure like `GCPUContext` is a possible ABI compatibility. While we can extend the information we pass through it, the ABI won't change. In your case it changes already

Reply:
...we don't strictly follow the ABI compatibility now, it is mostly FYI why use contexts instead of actual arguments.

Reply:
...we don't strictly follow the ABI compatibility now, it is mostly FYI why use contexts instead of actual arguments.

Reply:
Fixed.